Framework/include/Jet.h

```diff
@@ -5,90 +5,177 @@ class Jet
         {
-            if( !jetsID.at(ijet) ) continue;
-
             TLorentzVector lv(Jets.at(ijet));
             if( abs(lv.Eta()) < etaCut && 
                 lv.Pt() > 30 
                 )
             {
                 jets_pt30_->push_back(lv);
-                if(lv.Pt() > 45)
-                    jets_pt45_->push_back(lv);
+                
+                if(lv.Pt() > 40) {
+                    
+                    jets_pt40_->push_back(lv);
+                    
+                    if(lv.Pt() > 45) {   
+                        jets_pt45_->push_back(lv);
+                        jets_pt45_csv_->push_back(Jets_CSV.at(ijet));
+                    }
+                }
             }
         }
 
-        jets_pt30_clean_ = new std::vector<TLorentzVector>();
-        jets_pt45_clean_ = new std::vector<TLorentzVector>();
+        tr.registerDerivedVar("NJets",       Jets.size());
+        tr.registerDerivedVec("Jets_pt30",   jets_pt30_);
+        tr.registerDerivedVar("NJets_pt30", (jets_pt30_==nullptr)?0:jets_pt30_->size());
+        tr.registerDerivedVec("Jets_pt40",   jets_pt40_);
+        tr.registerDerivedVar("NJets_pt40", (jets_pt40_==nullptr)?0:jets_pt40_->size());
+        tr.registerDerivedVec("Jets_pt45",   jets_pt45_);
+        tr.registerDerivedVec("Jets_pt45_CSV",   jets_pt45_csv_);
+        tr.registerDerivedVar("NJets_pt45", (jets_pt45_==nullptr)?0:jets_pt45_->size());
 
-        if( cleanJets.size() != 0 ) {
-            for (unsigned int icjet = 0; icjet < cleanJets.size(); ++icjet)
-            {
-                if( !cleanJetsID.at(icjet) ) continue;
+        //Adding code to create a vector of GoodJets -> defined as the jet collection that eliminates the closest jet to any good lepton (muon or electron) if that delta R is less than 0.4 and the pT of the jet and lepton is approximately the same
+        
+        goodJets_       = new std::vector<TLorentzVector>();
+        goodjets_pt30_  = new std::vector<TLorentzVector>();
+        goodjets_pt40_  = new std::vector<TLorentzVector>();
+        goodjets_pt45_  = new std::vector<TLorentzVector>();
+        goodjets_CSV_   = new std::vector<double>();
+
+        std::vector<TLorentzVector>   goodjets_      = Jets;
+        std::vector<double>           goodjets_csv_  = Jets_CSV;
+        std::vector<int>              removedJetsIndices;
 
-                TLorentzVector lv(cleanJets.at(icjet));
-                if( abs(lv.Eta()) < etaCut &&
-                    lv.Pt() > 30
-                    )
-                {
-                    jets_pt30_clean_->push_back(lv);
-                    if( lv.Pt() > 45 )
-                        jets_pt45_clean_->push_back(lv);
+        if( NMuons > 0 ) {
+
+            for( TLorentzVector myMuon : Muons ) {
+                
+                double          tempDeltaR = 10.0;
+                int             tempJetIt  = -1;
+                
+                for( int myJetIt = 0; myJetIt < Jets.size(); ++myJetIt ) {
```

Comment:
Can we try to loop over the Jets collection once. Looks we do it three times in this file?

Reply:
Fixed everything, but still thinking on how to fix the looping over jets collection 3 times. The difficulty comes in that the GoodLeptons aren't defined until during the common variables.h, which necessarily has to be after Jet.h. otherwise, I can shrink it to 2 loops (one for the general Pt cut, and one for the Jet cleaning). 

Given that, if we move to clean jet collection, then we can eliminate one of the jet loops.